extensions/data-transfer/portability-data-transfer-flickr/src/main/java/org/dataportabilityproject/datatransfer/flickr/photos/FlickrPhotosImporter.java

```diff
@@ -80,46 +83,64 @@ public ImportResult importItem(UUID jobId, AuthData authData, PhotosContainerRes
     }
     RequestContext.getRequestContext().setAuth(auth);
 
-    // Store any album data in the cache because Flickr only allows you to create an album with a
-    // photo in it, so we have to wait for the first photo to create the album
     TempPhotosData tempPhotosData = jobStore.findData(TempPhotosData.class, jobId);
     if (tempPhotosData == null) {
       tempPhotosData = new TempPhotosData(jobId);
       jobStore.create(jobId, tempPhotosData);
     }
 
-    for (PhotoAlbum album : data.getAlbums()) {
-      tempPhotosData.addAlbum(CACHE_ALBUM_METADATA_PREFIX + album.getId(), album);
+    Preconditions.checkArgument(
+        data.getAlbums() != null || data.getPhotos() != null,
+        "" + "Error: There is no data to import");
+
+    if (data.getAlbums() != null) {
+      importAlbums(data.getAlbums(), tempPhotosData);
+      jobStore.update(jobId, tempPhotosData);
     }
-    jobStore.update(jobId, tempPhotosData);
-
-    for (PhotoModel photo : data.getPhotos()) {
-      try {
-        String photoId = uploadPhoto(photo);
-        String oldAlbumId = photo.getAlbumId();
-        TempPhotosData tempData = jobStore.findData(TempPhotosData.class, jobId);
-        String newAlbumId = tempData.lookupNewAlbumId(oldAlbumId);
-        if (Strings.isNullOrEmpty(newAlbumId)) {
-          // This means that we havent created the new album yet, create the photoset
-          PhotoAlbum album = tempData.lookupAlbum(CACHE_ALBUM_METADATA_PREFIX + oldAlbumId);
-          Photoset photoset =
-              photosetsInterface.create(
-                  COPY_PREFIX + album.getName(), album.getDescription(), photoId);
-          tempData.addAlbumId(oldAlbumId, photoset.getId());
-        } else {
-          // We've already created a new album, add the photo to the new album
-          photosetsInterface.addPhoto(newAlbumId, photoId);
+
+    if (data.getPhotos() != null) {
+      for(PhotoModel photo : data.getPhotos()) {
+        try {
+          importSinglePhoto(jobId, photo);
+        } catch (FlickrException | IOException e) {
+          return new ImportResult(ResultType.ERROR, "Error importing photo " + e.getMessage());
         }
-        jobStore.update(jobId, tempData);
-      } catch (FlickrException | IOException e) {
-        // TODO: figure out retries
-        return new ImportResult(
-            ImportResult.ResultType.ERROR, "Error importing item: " + e.getMessage());
       }
     }
+
     return new ImportResult(ImportResult.ResultType.OK);
   }
 
+  // Store any album data in the cache because Flickr only allows you to create an album with a
+  // photo in it, so we have to wait for the first photo to create the album
+  private void importAlbums(Collection<PhotoAlbum> albums, TempPhotosData tempPhotosData) {
+    for (PhotoAlbum album : albums) {
+      tempPhotosData.addAlbum(CACHE_ALBUM_METADATA_PREFIX + album.getId(), album);
```

Comment:
I ran into this with gmail.  We may need to store information prior to the actual mapping. I think it's ok if you want to add another method to tempPhotosData as it will be a messy object. Maybe addTempAlbum? in addition to addAlbumMapping()

Reply:
I renamed this addAlbum method to be the addTempAlbum - is that what you mean?